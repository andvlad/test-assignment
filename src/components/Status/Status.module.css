.root {
  padding: 4px;
  border-radius: 4px;
  text-transform: uppercase;
  letter-spacing: 2px;
  font-size: 14px;
  font-weight: 600;
  color: #404040;
}

.placed {
  background-color: lightgray;
}

.preparing,
.prepared,
.onTheWay {
  background-color: lightblue;
}

.delivered {
  background-color: lightgreen;
}
