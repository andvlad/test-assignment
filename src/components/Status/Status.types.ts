export interface IStatus {
  value: number;
  className?: string;
}
