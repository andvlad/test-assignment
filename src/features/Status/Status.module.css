.root {
  display: flex;
  margin-bottom: 25px;
}

a {
  cursor: pointer;
  color: #0066CC;
  text-decoration: none;
}

.statusComponent {
  margin: 0 8px 0 8px;
}

.okButton {
  margin: 0 8px 0 8px;
  background: none;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.okButton:hover,
.okButton:focus {
  background-color: #e0e0e0;
}
