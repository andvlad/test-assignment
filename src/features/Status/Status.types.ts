export interface IStatus {
  value: number;
}
