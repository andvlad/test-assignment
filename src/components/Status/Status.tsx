import React from 'react';

import { STATUSES } from 'app/constants';
import { IStatus } from './Status.types';
import styles from './Status.module.css';

export function Status({
  value,
  className,
}: IStatus) {
  const statusStyles = [
    styles.placed,
    styles.preparing,
    styles.prepared,
    styles.onTheWay,
    styles.delivered,
  ];

  return (
    <div className={`${styles.root} ${statusStyles[value]} ${className}`}>
      {STATUSES[value]}
    </div>
  );
}
