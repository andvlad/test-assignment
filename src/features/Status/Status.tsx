import React, {
  useCallback,
  useState,
  FormEvent,
} from 'react';

import { queryClient } from 'App';
import { STATUSES, QUERY_KEYS } from 'app/constants';
import { updateOrder } from 'services/order';
import { Status as StatusComponent } from 'components';
import { IStatus } from './Status.types';
import styles from './Status.module.css';

export function Status({
  value,
}: IStatus) {
  const [isStatusChangeActive, setIsStatusChangeActive] = useState<boolean>(false);
  const [selectValue, setSelectValue] = useState<number>(value);

  const onStatusChangeCallback = useCallback((event: FormEvent<HTMLAnchorElement>) => {
    event.preventDefault();

    setIsStatusChangeActive(!isStatusChangeActive);
  }, [isStatusChangeActive, setIsStatusChangeActive])

  const onSelectChangeCallback = useCallback((event: FormEvent<HTMLSelectElement>) => {
    event.preventDefault();

    setSelectValue(Number(event.currentTarget.value));
  }, [setSelectValue])

  const onOkClickCallback = useCallback(async (event: FormEvent<HTMLButtonElement>) => {
    event.preventDefault();

    if (selectValue === value) {
      return;
    }

    const order = queryClient.getQueryData<any>(QUERY_KEYS.ORDER);
    const now = new Date().toISOString();

    try {
      await updateOrder({
        ...order,
        status: selectValue,
        _updated: now,
        statusHistory: [
          ...order.statusHistory,
          {
            status: selectValue,
            timeStamp: now,
          }
        ],
      });
    } catch (e) {
      console.error(e)
    }

    setIsStatusChangeActive(false);
  }, [selectValue, value])

  return (
    <div className={styles.root}>
      Status:
      <StatusComponent
        value={value}
        className={styles.statusComponent}
      />
      {isStatusChangeActive ? (
        <>
          <select
            value={selectValue}
            onChange={onSelectChangeCallback}
          >
            {STATUSES.map((status, index) => (
              <option
                key={status}
                value={index}
              >
                {status}
              </option>
            ))}
          </select>
          <button
            onClick={onOkClickCallback}
            className={styles.okButton}
          >
            Ok
          </button>
          <a onClick={onStatusChangeCallback}>Cancel</a>
        </>
      ) : (
        <a onClick={onStatusChangeCallback}>Change status</a>
      )}
    </div>
  );
}
