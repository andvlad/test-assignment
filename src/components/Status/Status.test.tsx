import React from 'react';
import { render } from '@testing-library/react';

import { Status } from './Status';

it('matches snapshot', () => {
  const { asFragment } = render(
    <Status value={1} />
  );

  expect(asFragment()).toMatchSnapshot();
});
