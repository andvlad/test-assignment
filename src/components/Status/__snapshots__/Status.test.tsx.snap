// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="root preparing undefined"
  >
    Preparing
  </div>
</DocumentFragment>
`;
